RFCs/0051-exclude-operator.adoc

```diff
@@ -0,0 +1,1700 @@
+                                CASE WHEN v_2 IS STRUCT THEN (
+                                    PIVOT v_3 AT attr_3
+                                    FROM UNPIVOT v_2 AS v_3 AT attr_3
+                                    WHERE LOWER(attr_3) NOT IN [LOWER('bar')]
+                                )
+                                ELSE v_2
+                                END
+                            FROM v_1 AS v_2 AT idx_2
+                            ORDER BY idx_2
+                        )
+                        -- WHEN v_1 IS BAG THEN ... 
+                        -- same as for LIST but remove `AT` and `ORDER BY`
+                        ELSE v_1
+                        END
+                    ELSE v_1
+                    END
+                ) AT attr_1 FROM UNPIVOT t AS v_1 AT attr_1
+            )
+            ELSE t
+            END
+    }
+    FROM 
+    <<
+        {'a': [{'foo': 0, 'bar': 1, 'baz': 2}, {'foo': 3, 'bar': 4, 'baz': 5}]},
+        {'a': {'foo': 6, 'bar': 7, 'baz': 8}},
+        {'a': {'a1': {'foo': 9, 'bar': 10, 'baz': 11}, 'a2': {'foo': 12, 'bar': 13, 'baz': 14}}}
+    >> AS t
+)
+----
+
+Output:
+[source,partiql,subs="+{markup-in-source}"]
+----
+<<
+  {
+    'a': [
+      {
+        'foo': 0,
+        'baz': 2
+      },
+      {
+        'foo': 3,
+        'baz': 5
+      }
+    ]
+  },
+  {
+    'a': {
+      'foo': 6,
+      'baz': 8
+    }
+  },
+  {
+    'a': {
+      'a1': {
+        'foo': 9,
+        'baz': 11
+      },
+      'a2': {
+        'foo': 12,
+        'baz': 14
+      }
+    }
+  }
+>>
+----
+
+== Drawbacks
+
+`EXCLUDE` (or similar clause) is not part of the SQL or SQL++ standard. If `EXCLUDE` is added in a future standard, it's possible the syntax and semantics may change.
+
+== Rationale and alternatives
+[qanda]
+In the original spec issue (https://github.com/partiql/partiql-spec/issues/39[partiql-spec#39]), `EXCEPT` was included as the keyword for this clause. Why was the keyword `EXCLUDE` chosen?::
+
+`EXCLUDE` was chosen over `EXCEPT` since `EXCEPT` could be confused with the set/bag operator `EXCEPT`. `EXCLUDE` was also chosen by the SQL++ implementation, AsterixDB, through some similar reasoning:
++
+[quote,https://issues.apache.org/jira/browse/ASTERIXDB-3059]
+____
+'EXCLUDE' (used in lieu of 'EXCEPT' to avoid confusion with the set operation)
+____
++
+Also of the databases sampled that have a similar clause (see <<Prior art>>), more had chosen `EXCLUDE` over `EXCEPT`.
+
+Why is `EXCLUDE` modeled as a binding tuple operator as opposed to a value expression?::
+
+We had also considered modeling `EXCLUDE` as a value operation evaluated after the `<select clause>`. Evaluating `EXCLUDE` last could contradict the PartiQL specification's assertion that the `<select clause>` is evaluated last, which may add confusion. There were also some additional edge cases that complicated defining `EXCLUDE` as a value operator. For example, let's look at the following query:
```

Comment:
> Did you explore `EXCLUDE` as a value operation (maybe just a special-form) evaluated during the `<select clause>`? Similar to how any other special-form function invocation would work in the projection list?

Could you clarify what you mean by special-form function? Is this referring to non-standard function syntax (kinda like what substring, overlay do)?

Regarding modeling `EXCLUDE` as a value operation/function evaluated within the `SELECT` clause, we had also explored this option. These aren't quite functions though, since function arguments expect value expressions and exclude paths do not return values.

The above way you had modeled expressions evaluated left-to-right could be problematic. Consider `SELECT EXCLUDE t.a[0], EXCLUDE t.a[1].field FROM ...`. Should the second exclude path remove `field` from the `a`'s original `1` index or after evaluating the first exclude path (i.e. originally the `2` index). We found considering the paths as associative and evaluating the paths in parallel (i.e. applying all exclude paths on the same value/binding) as more intuitive.

Reply:
1. Yes, meaning non-standard function syntax.
2. Yes -- but that is the AST, which isn't the same as execution. We don't operate on the AST -- we convert `TRIM` to `$trim_leading` and `CURRENT_USER` to `$current_user`. The SQL:1999 Specification even talks about the conversion of the syntactic `BETWEEN` to comparison operators -- the theoretical `between` AST node disappears and gets replaced. We could just as easily say `<lhs> EXCLUDE ( <exclude-arg-1>, <exclude-arg-2> )` gets converted to `EXCLUDE(<lhs>, [ <exclude-arg-1-stringified>, <exclude-arg-2-stringified>])`. Anyways, it *could* provide flexibility, and I believe it's possible to model.
3. We could probably look at BigQuery which expects the RHS to be parenthesized. Ex: `SELECT * EXCEPT (a, b)`. I would imagine this query in BigQuery strips out both `a` and `b` simultaneously.

Given PartiQL's flexibility, I'm personally intrigued by the prospect of writing:
```
SELECT VALUE {
  'full_name': person.l_name || ', ' person.f_name,
  'encrypted_ssn': udf_to_encrypt(person.ssn),
  'all_other_details': person EXCLUDE (f_name, l_name, ssn)
} FROM Persons AS person
```

Especially since tuples are first-class citizens, I find it long overdue for some practical built-ins (beyond `TUPLEUNION` which PLK doesn't have implemented -- yet). It seems like a potential opportunity to allow for its introduction, especially since SQL's `SELECT` gets transformed to PartiQL's `SELECT VALUE <tuple>`.

Reply:
Cannot be a value operator if you want to ever prune binding tuple variables.